client/adjudicate.go

```diff
@@ -417,23 +417,34 @@ func (c *Channel) ensureRegistered(ctx context.Context) error {
 		return nil
 	}
 
-	registeredEvents := make(chan *channel.RegisteredEvent)
-
+	events := make(chan channel.AdjudicatorEvent)
 	// Start event subscription.
 	sub, err := c.adjudicator.Subscribe(ctx, c.Params().ID())
 	if err != nil {
 		return errors.WithMessage(err, "subscribing to adjudicator events")
 	}
-	go waitForRegisteredEvent(sub, registeredEvents)
+	defer func() {
+		if err := sub.Close(); err != nil {
+			c.Log().Warn("Subscription closed with error:", err)
+		}
+	}()
+	go awaitRegistered(sub, events)
 
 	// Register.
 	err = c.registerDispute(ctx)
 	if err != nil {
-		return errors.WithMessage(err, "registering")
+		// Only log because channel may already be registered.
+		c.Log().Warnf("registering: %v", err)
 	}
 
 	select {
-	case e := <-registeredEvents:
+	case e := <-events:
+		// We set phase registered to ensure that we are in the correct phase
+		// even if the channel has been registered by someone else.
+		err = c.setRegisteredRecursive(ctx)
```

Comment:
Is this function idempotent? Would it break when there are multiple events received here?

Reply:
Yes, it's fine to call it twice.